src/StreamJsonRpc/Exceptions/LocalRpcException.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace StreamJsonRpc
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Reflection;
+    using System.Text;
+    using Microsoft;
+    using StreamJsonRpc.Protocol;
+
+    /// <summary>
+    /// An exception that may be thrown within a locally invoked server method, and carries with it data that influences the JSON-RPC error message's error object.
+    /// </summary>
+    public class LocalRpcException : Exception
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="LocalRpcException"/> class.
+        /// </summary>
+        public LocalRpcException()
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="LocalRpcException"/> class.
+        /// </summary>
+        /// <param name="message">The error message that explains the reason for the exception.</param>
+        public LocalRpcException(string message)
+            : base(message)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="LocalRpcException"/> class.
+        /// </summary>
+        /// <param name="message">The error message that explains the reason for the exception.</param>
+        /// <param name="inner">The exception that is the cause of the current exception, or a null reference (Nothing in Visual Basic) if no inner exception is specified.</param>
+        public LocalRpcException(string message, Exception inner)
+                : base(message, inner)
+        {
+        }
+
+        /// <summary>
+        /// Gets or sets the value for the error.data property.
+        /// </summary>
+        public object ErrorData { get; set; }
+
+        /// <summary>
+        /// Gets or sets the value for the error.code property.
+        /// </summary>
+        /// <value>Any value, excluding the [-32768, -32000] range, which is reserved.</value>
```

Comment:
Is the reason you don't validate that these aren't set because other code uses this and does set that value range appropriately? Perhaps having this version validate and then override error code setter in a derived class to allow only the reserved range to be set.

Reply:
I should update this doc, since it suggests that the value would never be in that reserved range. It certainly can be, but apps shouldn't be *setting* within that range.
As for enforcing by code that no one ever use that range, I'm leaning toward thinking that's overkill. Particularly since whatever might set that odd error code is an exception path for the app and probably poorly tested, I wouldn't want to hide whatever the original error was with a new error.